mcf-browser/src/back-end/parse-tmcf.ts

```diff
@@ -174,6 +185,128 @@ class ParseTmcf {
+   *     Each CsvRow element of the array represents one row of the csv.
+   * @return {ParsedCsvRow | null} an object containing the facet, date,
+   *     and value
+   */
+  getFacetAndValueFromRow(template: string, row: CsvRow)
+  : ParsedCsvRow | null {
+    type Properties = {
+      [propertyLabel: string]: string | undefined
+    }
+    const properties: Properties = {};
+
+    // Parse row
+    for (const line of template.split('\n')) {
+      if (!line.trim() || !shouldReadLine(line)) {
+        continue;
+      }
+
+      const propLabel = line.split(':')[0].trim();
+      const propValues = line.substring(line.indexOf(':') + 1).trim();
+
+      if (propLabel === 'Node') {
+        if (propValues.includes(',')) {
+          throw new Error('cannot have multiple ids for Node declaration');
+        }
+        const entityID = getEntityID(propValues);
+        if (entityID) {
+          properties[propLabel] =
+              this.getLocalIdFromEntityId(entityID) as string;
+        } else {
+          properties[propLabel] = propValues;
+        }
+      } else {
+        const filledValues = this.fillPropertyValues(propValues, row);
+        properties[propLabel] = filledValues;
+      }
+    }
+
+    // Ignore non-StatVarObservations
+    if (!properties.typeOf || properties.typeOf !== 'dcs:StatVarObservation') {
+      return null;
+    }
+
+    // Generate output
+    const facet = Series.toID(
+        properties.variableMeasured,
+        properties.observationAbout,
+        properties.provenance,
+        properties.measurementMethod,
+        properties.observationPeriod,
+        properties.unit,
+        properties.scalingFactor ?
+          parseFloat(properties.scalingFactor) : undefined,
+    );
+    const date = properties.observationDate ? properties.observationDate : '';
+    const value = properties.value ? properties.value : '';
+    return {facet, date, value};
+  }
+
+  /**
+   * Creates a mapping from facet to values from a string representation of
+   * TMCF file and the json representation of a CSV file. The whole
+   * template from the tmcf is populated with values for each row of the csv.
+   * @param {string} template The string representation of a tmcf file.
+   * @param {Array<CsvRow>} csvRows The json representation of the csv file.
+   *     Each CsvRow element of the array represents one row of the csv.
+   * @return {TimeDataObject} The generated mcf as an Object.
+   */
+  csvToDataPoint(template: string, csvRows: CsvRow[]) : TimeDataObject {
+    this.csvIndex = 1;
+
+    const datapoints: TimeDataObject = {};
+    for (const row of csvRows) {
+      const parsedCsvRow = this.getFacetAndValueFromRow(
+          template,
+          row,
+      );
+
+      if (parsedCsvRow) {
+        const {facet, date, value} = parsedCsvRow;
+        const parsedValue = value !== '' ? parseFloat(value) : undefined;
+        datapoints[facet] = datapoints[facet] ?
```

Comment:
nit: maybe just this instead?
```
if (!(facet in datapoints)) {
datapoints[facet] = {};
}
datapoints[facet][date] = parsedValue;
```

Reply:
I tried this earlier but TS throws an error for me that `datapoints[facet]` could be undefined since the type was defined with index signatures. Even when I wrap it with an if statement check ( i.e. `if (datapoints[facet]) ...`  ), it still gives me that error

Reply:
oh interesting, ok, that's fine then